lib/QoreFtpClient.cpp

```diff
@@ -91,247 +91,270 @@ class FtpResp {
-               if (p > (start + 3)) {
-                  // if we got the whole response
-                  if (isdigit(*start) && isdigit(start[1]) && isdigit(start[2]) && start[3] == ' ') {
-                     code = ((*start - 48) * 100) + ((start[1] - 48) * 10) + start[2] - 48;
-                     // if we read more data, then store it in the buffer
-                     if (p[1] != '\0') {
-                        buffer.set(&p[1]);
-                        resp->terminate(p - resp->getBuffer() + 1);
-                     }
-                     break;
-                  }
-               }
-               start = p + 1;
+        }
+
+        // verify protocol
+        if (url.getProtocol()) {
+            if (!url.getProtocol()->compare("ftps"))
+                secure = secure_data = true;
+            else if (url.getProtocol()->compare("ftp")) {
+                xsink->raiseException("UNSUPPORTED-PROTOCOL", "'%s' not supported (expected 'ftp' or 'ftps')", url.getProtocol()->getBuffer());
+                return;
             }
-            // if we have not got the whole message
-            else if (*p == '\0') {
-               QoreStringNodeHolder r(control.recv(timeout_ms, xsink));
-               if (*xsink) {
-                  disconnectIntern();
-                  return 0;
-               }
-               if (!r) {
-                  disconnectIntern();
-                  xsink->raiseException("FTP-RECEIVE-ERROR", "short message received on control port");
-                  return 0;
-               }
-               //printd(FTPDEBUG, "QoreFtpClient::getResponse() read %s\n", r->getBuffer());
-               // in case the buffer gets reallocated
-               int pos = p - resp->getBuffer();
-               // cannot maintain start across buffer reallocations
-               size_t offset = p - start;
-               resp->concat(*r);
-               p = resp->getBuffer() + pos;
-               start = p + offset;
+        }
+
+        // set username
+        user = url.take_username();
+        // set password
+        pass = url.take_password();
+        // set host
+        host = url.take_host();
+        port = url.getPort() ? url.getPort() : DEFAULT_FTP_CONTROL_PORT;
+    }
+
+    DLLLOCAL QoreHashNode* getControlPeerInfo(ExceptionSink* xsink, bool host_lookup) const {
+        AutoLocker al(m);
+        return control.getPeerInfo(xsink, host_lookup);
+    }
+
+    DLLLOCAL QoreHashNode* getDataPeerInfo(ExceptionSink* xsink, bool host_lookup) const {
+        AutoLocker al(m);
+        return data.getPeerInfo(xsink, host_lookup);
+    }
+
+    DLLLOCAL QoreHashNode* getControlSocketInfo(ExceptionSink* xsink, bool host_lookup) const {
+        AutoLocker al(m);
+        return control.getSocketInfo(xsink, host_lookup);
+    }
+
+    DLLLOCAL QoreHashNode* getDataSocketInfo(ExceptionSink* xsink, bool host_lookup) const {
+        AutoLocker al(m);
+        return control.getSocketInfo(xsink, host_lookup);
```

Comment:
shouldn't getDataSocketInfo actually call `data.getSocketInfo` instead of `control.get...`